include/reader_writer_lock.hpp

```diff
@@ -0,0 +1,110 @@
+#ifndef READER_WRITER_LOCK_HPP
+#define READER_WRITER_LOCK_HPP
+
+#include <atomic>
+#include <cstdint>
+
+namespace fair::graph {
+
+enum class ReaderWriterLockType { READ, WRITE };
+
+/**
+ * @brief ReaderWriterLock is multi-reader-multi-writer atomic lock meant to protect a resource
+ * in situations where the thread is not allowed to block.
+ *
+ * The lock is implemented using atomic CAS-loops on a counter, which is
+ * incremented (/decremented) when a thread acquires a read (/write) lock, and
+ * decremented (/incremented) when the thread releases the read (/write) lock.
+ *
+ * N.B. The lock is unlocked when the counter reaches 0.
+ */
+class ReaderWriterLock {
+    alignas(64) mutable std::atomic<std::int64_t> _activeReaderCount{ 0 };
```

Comment:
re-use the destructive interference or `fair::meta::kCacheLine` short-hand definition.

Reply:
Good catch!